internal/mining/bgblktmplgenerator.go

```diff
@@ -129,6 +129,48 @@ type regenEvent struct {
 	value  interface{}
 }
 
+// waitGroup behaves simlarly to a sync.WaitGroup without the restriction that
+// Adds() and Waits() must be synchronized if the wait group is empty.
+type waitGroup struct {
+	mtx sync.Mutex
+	c   int64
+	dc  chan struct{}
+}
+
+func (wg *waitGroup) Add(i int64) {
+	wg.mtx.Lock()
+	wg.c += i
+	switch {
+	case wg.c < 0:
+		panic("counter cannot be negative")
+
+	case wg.c > 0 && wg.dc == nil:
+		// First time the counter is increased, create the doneChan.
```

Comment:
this isn't necessarily the first time the counter is increased, as the channel is assigned nil when it reaches a zero count again.

(perhaps this could cause some data or logic races when the waitgroup unblocks, before all jobs have been added and finished?)

Reply:
"First time" was meant as "First time since it was last empty" (not necessarily first time ever during the wait group's lifetime). Will try to improve the comment.

> (perhaps this could cause some data or logic races when the waitgroup unblocks, before all jobs have been added and finished?)

Not completely sure the workflow you mean here, but if you mean something like:
```
| Goroutine 1 | Goroutine 2   | Goroutine 3 |
|   Add(1)    |               |             |  Time 
|             |     Wait()    |             |  ||
|   Done()    |(wait unblocks)|             |  \/
|             |               |    Add(1)   |   
|             |               |             |
|             |               |    Done()   |
```

Then yes, Goroutine 2 unblocks before Goroutine 3 had a chance to call `Add(1)` (which is going to create a new signalling chan, eventually closed by Goroutine 3's `Done()` call).

However, the critical issue for the usage pattern for this particular wait group in the background template generator is that the Goroutines themselves are unsynchronized: there's no way to know the future Add() will even happen because it depends on external factors (for example, a block coming in or a vote) and there's no way to know the Wait() will happen (because it comes on a Goroutine started by an rpc handler).

The bg template generator needs to ensure that while there are _known_ events that should delay template generation, any clients (i.e. calls to Wait()) are paused. However, after the last known one is processed (i.e. last call do Done() which closes the signalling chan) releases all current waiters it can't guarantee that a _new_ event won't arrive in the next millisecond or whatever (because that event is arbitrary).